### PR TITLE
[Keyboard] Add ADM42 Keyboard

### PR DESCRIPTION
**Pull request by lleonini**
## Description

Add support for the ADM42 keyboard.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/adm42/config.h [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+#define MATRIX_COLS 12
+
+#define DIODE_DIRECTION ROW2COL
+
+#define HOLD_ON_OTHER_KEY_PRESS_PER_KEY
+#define TAPPING_FORCE_HOLD_PER_KEY
+#define IGNORE_MOD_TAP_INTERRUPT_PER_KEY
+#define TAPPING_TERM 200
+
+#define USB_POLLING_INTERVAL_MS 1
```
Comment (drashna) [inline keyboards/adm42/config.h:39]: These are the default values and don't need to be specified.

```suggestion
```

File: keyboards/adm42/config.h [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+// Memory saver
+#define NO_ACTION_ONE_SHOT
+#if !defined(CONSOLE_ENABLE)
+	#if !defined(NO_DEBUG)
+		#define NO_DEBUG
+	#endif
+	#if !defined(NO_PRINT)
+		#define NO_PRINT
+	#endif
+#endif
```
Comment (drashna) [inline keyboards/adm42/config.h:56]: If console isn't enabled these are automatically defined.  If console is enabled, you don't want to disable these. 
```suggestion
```

File: keyboards/adm42/readme.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
+# ADM42
+
+![ADM42](https://adm42.dev/images/adm42.jpg)
```
Comment (drashna) [inline keyboards/adm42/readme.md:3]: please rehost the image on Imgur or similar image hosting site.

File: keyboards/adm42/rev4/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+        {132, 36}, {148, 34}, {164, 29}, {180, 28}, {198, 34}, {215, 32},
+        {212, 22}, {196, 23}, {178, 17}, {161, 18}, {146, 24}, {130, 25},
+        {126, 14}, {143, 13}, {159, 8},  {175, 6},  {194, 13}, {210, 11}
+    }, {
+        4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
+        4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
+        4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
+        4, 4, 4, 4, 4, 4
+    }
+};
```
Comment (drashna) [inline keyboards/adm42/rev4/keymaps/default/keymap.c:39]: This should be moved to the keyboard's c file, not the keymap.

```suggestion
```

File: keyboards/adm42/rev4/keymaps/default/keymap.c [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+
+#define _QWERTY 0
+#define _COLEMAKDH 1
+#define _SPECIAL 2
+#define _EXTRA 3
+#define _ADM 4
+#define _SETUP 5
```
Comment (drashna) [inline keyboards/adm42/rev4/keymaps/default/keymap.c:46]:
```suggestion
enum custom_layers {
    _QWERTY,
    _COLEMAKDH,
    _SPECIAL,
    _EXTRA,
    _ADM,
    _SETUP,
};
```

File: keyboards/adm42/config.h [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+#define PRODUCT         ADM42
+
+#define MATRIX_ROWS 4
+#define MATRIX_COLS 12
+
+#define DIODE_DIRECTION ROW2COL
+
+#define HOLD_ON_OTHER_KEY_PRESS_PER_KEY
+#define TAPPING_FORCE_HOLD_PER_KEY
+#define IGNORE_MOD_TAP_INTERRUPT_PER_KEY
```
Comment (drashna) [inline keyboards/adm42/config.h:36]: These should be defined at the keymap/user level, not kb level. 
```suggestion
```

File: keyboards/adm42/rules.mk [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+COMMAND_ENABLE = no         # Commands for debug and configuration
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+BLUETOOTH_ENABLE = no       # Enable Bluetooth
+AUDIO_ENABLE = no           # Audio output
+
+DEBOUNCE_TYPE = sym_eager_pk
+
+EXTRAFLAGS += -flto
```
Comment (drashna) [inline keyboards/adm42/rules.mk:18]:
```suggestion
LTO_ENABLE = yes
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by lleonini** [comment]
Thanks for your feedback, I have commited the changes. Should I create a new pull request ?

**Review by drashna — APPROVED** [review]

**Comment by drashna** [comment]
> Thanks for your feedback, I have commited the changes. Should I create a new pull request ?

No need.  The suggestions commit directly to the branch used for the PR, and when the branch is updated, the PR is updated. 

So there is no need to re-open, since the PR gets updated.

File: keyboards/adm42/rules.mk [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+COMMAND_ENABLE = no         # Commands for debug and configuration
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+BLUETOOTH_ENABLE = no       # Enable Bluetooth
+AUDIO_ENABLE = no           # Audio output
+
+DEBOUNCE_TYPE = sym_eager_pk
+
+LTO_ENABLE = yes
```
Comment (fauxpark) [inline keyboards/adm42/rules.mk:18]:
```suggestion
# MCU name
MCU = atmega32u4

# Bootloader selection
BOOTLOADER = atmel-dfu

# Build Options
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
MOUSEKEY_ENABLE = no        # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
NKRO_ENABLE = no            # Enable N-Key Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
AUDIO_ENABLE = no           # Audio output
DEBOUNCE_TYPE = sym_eager_pk
LTO_ENABLE = yes
```

File: keyboards/adm42/readme.md [hunk 9]
```diff
@@ -0,0 +1,27 @@
+# ADM42
+
+![ADM42](https://i.imgur.com/pviCQSQ.jpeg)
```
Comment (fauxpark) [inline keyboards/adm42/readme.md:3]:
```suggestion
![ADM42](https://i.imgur.com/pviCQSQh.jpeg)
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — APPROVED** [review]

**Comment by drashna** [comment]
Thanks!

**Pull request merged**